ui/src/components/drafts/form/themes/grommet/widgets/RegExInput.js

```diff
@@ -0,0 +1,583 @@
+  OK: [<Status size="small" value="unknown" />, ""]
+};
+
+function minV(minVal) {
+  return val => Math.max(minVal, val);
+}
+
+export function hashStr(str) {
+  let hashVal = 5381,
+    i = str.length;
+
+  while (i) hashVal = (hashVal * 33) ^ str.charCodeAt(--i);
+  return (hashVal >>> 0) + 12;
+}
+
+class RxInputBase extends Component {
+  constructor(props) {
+    super(props);
+    this.state = this.getInitialState();
+    this._onChange = this._onChange.bind(this);
+    this._onKeyDown = this._onKeyDown.bind(this);
+    this._onKeyPress = this._onKeyPress.bind(this);
+    this._onPaste = this._onPaste.bind(this);
+    this._onFocus = this._onFocus.bind(this);
+    this._onBlur = this._onBlur.bind(this);
+    this.input = null;
+  }
+  // propTypes: {
+  //   mask: React.PropTypes.object.isRequired,
+  //   name: React.PropTypes.string.isRequired,
+  //   popover: React.PropTypes.string,
+  //   selection: React.PropTypes.object,
+  //   value: React.PropTypes.string,
+  // },
+
+  // get input() {
+  // 	return this.myRef.current;
+  // }
+
+  getInitialState() {
+    let options = {
+      pattern: this.props.mask || /.*/,
+      value: this.props.value || ""
+    };
+    return {
+      focus: false,
+      value: this.props.value || "",
+      selection: this.props.selection,
+      mask: new RXInputMask(options)
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (this.props.mask.toString() !== nextProps.mask.toString()) {
+      //this.state.mask.setPattern(nextProps.mask, {value: this.state.mask.getRawValue()});
+      this.state.mask.setPattern(nextProps.mask, {
+        value: nextProps.value,
+        selection: this.state.mask.selection
+      });
+      this.setState({
+        selection: this.state.selection,
+        value: nextProps.value
+      });
+    } else if (this.props.value !== nextProps.value) {
+      this.state.mask.setValue(nextProps.value);
+    }
+  }
+
+  // static getDerivedStateFromProps(nextProps, state) {
+  // 	//if (this.props.mask.toString() !== nextProps.mask.toString()) {
+  //     let value = nextProps.value||'';
+  //     console.log('getDerivedStateFromProps',{oldValue: state.value, value, oldmask:state.mask.pattern.toString(),mask: nextProps.mask.toString() });
+  // 	if (state.mask.pattern.toString() !== nextProps.mask.toString()) {
+  // 		//this.state.mask.setPattern(nextProps.mask, {value: this.state.mask.getRawValue()});
+  // 		state.mask.setPattern(nextProps.mask, {value, selection: state.mask.selection});
+  // 		//this.setState({ selection: this.state.selection, value: nextProps.value});
+  // 		return {...state, value: nextProps.value};
+  // 	}
+  // 	else if (state.value !== value) {
+  // 		state.mask.setValue(nextProps.value);
+  // 		console.log("value change");
+  // 		return {...state };
+  // 	}
+  // 	return null;
+  // }
+
+  _updateMaskSelection() {
+    this.state.mask.selection = getSelection(this.input);
```

Comment:
We should not mutate the state directly. Better to use `setState`

Reply:
Let's leave as is for now